_posts/2024-10-14-performance-boost-windows.md

```diff
@@ -0,0 +1,390 @@
+
+
+## Vectorization Optimization in PyTorch 2.4.1 and later
+
+Prior to PyTorch 2.4.1, the Windows build of PyTorch lacked [SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) for vectorization optimizations, a feature that the Linux build leveraged for improved performance. This discrepancy was due to the [SLEEF](https://github.com/shibatch/sleef) Library’s integration issues on Windows, which is a SIMD Library for Evaluating Elementary Functions, vectorized libm and DFT and is essential for efficient trigonometric calculations. Through a collaborative effort with engineers from ARM and Qualcomm, these challenges were resolved, enabling the integration of SIMD into PyTorch for Windows. The PyTorch 2.4.1 update has thus significantly enhanced PyTorch’s CPU performance on Windows, as shown in Figure 2.1.
+
+
+![performance comparison chart](/assets/images/performance-boost-windows/fg4.png){:style="width:100%"}
+
+
+*PyTorch CPU Performance Improvement on Windows with Vertorization Optimization*
+
+*Figure 2.1: Relative throughput improvement achieved by upgrading from PyTorch CPU version 2.1.2 to 2.4.1 (higher is better).*
+
+As shown in the graph below, we see that PyTorch CPU performance on Windows ahieved the performance on Linux.
+
+
+![performance comparison chart](/assets/images/performance-boost-windows/fg5.png){:style="width:100%"}
+
+
+*Windows vs Linux Performance on PyTorch 2.4.1*
+
+*Figure 2.2: Relative performance of Windows vs Linux with PyTorch version 2.4.1 (higher is better).*
+
+
+## CONCLUSION
+
+From PyTorch 2.0.1 to PyTorch 2.4.1, the CPU performance gap between Windows and Linux has been continuously narrowing. We compared the ratio of CPU performance on Windows to CPU performance on Linux across different versions, and the results are shown in the following graph.
+
+
+![performance comparison chart](/assets/images/performance-boost-windows/fg6.png){:style="width:100%"}
+
+
+*Windows vs Linux Performance on different version of PyTorch*
+
+*Figure 3: Performance Ratio for Windows to Linux with different version of PyTorch (higher is better).*
+
+The graph shows that with PyTorch 2.4.1, CPU performance on Windows has nearly converged with that on Linux, and on some models, it has even surpassed Linux. For example, in the case of DistillBERT and RoBERTa models, the CPU performance ratio of Windows to Linux has achieved a remarkable 102%. However, certain models, including MobileNet-v3, still show a performance discrepancy. Intel engineers will continue to collaborate with Meta engineers, to reduce the performance gap of PyTorch CPU between Windows and Linux.
+
+
+## HOW TO TAKE ADVANTAGE OF THE OPTIMIZATIONS
+
+Install PyTorch CPU 2.4.1 on Windows from the [official repository](https://pytorch.org/get-started/locally/), and you may automatically experience a performance boost with memory allocation and vectorizations.

```

Comment:
Since PyTorch 2.5 is going to be released soon, should we replace "Install PyTorch CPU 2.4.1 " to "Install PyTorch CPU 2.4.1 or later"

Reply:
Ok, updated. Thanks.